cypress/plugins/index.ts

```diff
@@ -0,0 +1,242 @@
+import * as path from "path";
+import * as childProcess from "child_process";
+import fetch from "node-fetch";
+import * as util from "util";
+import waitOn from "wait-on";
+import * as uuid from "uuid";
+import * as fs from "fs-extra";
+
+const exec = util.promisify(childProcess.exec);
+
+const HOST = "127.0.0.1";
+const ROOT_PATH = path.join(__dirname, "../../");
+const CYPRESS_WORKSPACE_PATH = path.join(ROOT_PATH, "cypress/workspace");
+const PROXY_PATH = path.join(ROOT_PATH, "proxy/target/debug/radicle-proxy");
+
+let nodes: Node[] = [];
+
+const getNode = (id: NodeId) => {
+  const node = nodes.find(node => {
+    return node.id === id;
+  });
+
+  if (!node) {
+    throw `Could not find node by id ${id}`;
+  }
+
+  return node;
+};
+
+const sleep = async (ms: number) => {
+  await new Promise(resolve => setTimeout(resolve, ms));
+};
+
+type NodeId = number;
+
+interface Node {
+  id: NodeId;
+  httpListen: number;
+  peerListen: number;
+  workspacePath: string;
+  pid?: number;
+  authToken?: string;
+  peerAddress?: string;
+}
+
+export interface OnboardedNode {
+  authToken: string;
+  peerAddress: string;
+  workspacePath: string;
+}
+
+export default (
+  on: Cypress.PluginEvents,
+  _config: Cypress.PluginConfigOptions
+): void => {
+  on("task", {
+    // Kill any nodes that were started by this plugin.
+    async killAllNodes() {
+      console.log("  ### killAllNodes()");
+
+      nodes.forEach(async node => {
+        try {
+          await exec(`kill -9 ${node.pid}`);
+        } catch (err) {
+          console.log("    ERROR: ", err.message);
+        }
+      });
+      nodes = [];
+      return null;
+    },
+    // Start a new `radicle-proxy` node, the `id` is used as a port number for
+    // both API TCP connections as well as UDP peer connections.
+    //
+    // This task also sets up an empty workspace directory where we can create
+    // new project and checkout directories or store any other data necessary
+    // for a testcase:
+    //
+    //   ./radicle-upstream/cypress/workspace/<RANDOM_UUID>
+    //
+    // This command waits on the proxy process to start listening on its HTTP
+    // API port before returning.
+    async startNode(id: NodeId) {
+      console.log(`  ### startNode(${id})`);
+
+      const node: Node = {
+        id: id,
+        httpListen: id,
+        peerListen: id,
+        workspacePath: path.join(CYPRESS_WORKSPACE_PATH, uuid.v4()),
```

Comment:
With support for `RAD_HOME` landing in the proxy (#1527) we should use this we should set it to this directory. Otherwise it looks to me like the proxies will share state?

Reply:
They wouldn't share state because of the random UUID afaict, but incorporating `RAD_HOME` is a good shout regardless because one of our tests later should be about seeing new commits. Env variables can be per process, right? :thinking:

Reply:
> They wouldn't share state because of the random UUID

If I understand this code correctly a random UUID is only used in the workspace path but the workspace path itself is never used. But isolation is provided through the `--test` flag so maybe we just don’t need the workspace path.

Reply:
Mmmm good point, I actually didn't trace where this path was used!

Reply:
> With support for RAD_HOME landing in the proxy (#1527) we should use this we should set it to this directory. Otherwise it looks to me like the proxies will share state?

The intention of `CYPRESS_WORKSPACE_PATH` is to store anything that is needed for a test-case scoped to a running node. It's for storing newly created project or project checkout working directories, so that we can inspect the contents of those directories from within a test.

This is _not_ the path where we store the internal monorepo/keys/sled/electron things. The internal state is kept in a tempdir when we run tests via the `--test` flag.

We keep the `CYPRESS_WORKSPACE_PATH` within the `radicle-upstream` repository, so that it would work both locally as well as on CI.

Does this clear things up?

> If I understand this code correctly a random UUID is only used in the workspace path but the workspace path itself is never used. But isolation is provided through the --test flag so maybe we just don’t need the workspace path.

The workspace path is used in the test: https://github.com/radicle-dev/radicle-upstream/pull/1492/files#diff-c1135bec4b86c0f227a805cabfd0d9933c669329ede40a2ef36a2c031c066264R21-R22.

I wonder if there's a way to make this more clear, perhaps a comment?

Reply:
Thanks for the explanation @rudolfs. That clears things up.

As an alternative (since the workspace path and the node are not intrinsically linked) we could also prepare the temporary directory independently of the node in the test case. Similar to what we do in [other tests](https://github.com/radicle-dev/radicle-upstream/blob/929f4593b70a419a88141b26d65751a35c064768/cypress/integration/project/checkout.spec.ts#L5).

Reply:
Added the tempdir wrapper to our existing `commands.ts`: https://github.com/radicle-dev/radicle-upstream/pull/1492/commits/e46870ae78be463f7244faa979625f2a267cc92b.